hi-ml-histopathology/src/histopathology/models/deepmil.py

```diff
@@ -285,3 +272,68 @@ def test_epoch_end(self, epoch_results: EpochResultsType) -> None:  # type: igno
                 epoch_results=epoch_results,
                 is_global_rank_zero=self.global_rank == 0
             )
+
+
+class TilesDeepMILModule(BaseDeepMILModule):
+    """Base class for Tiles based deep multiple-instance learning."""
+
+    def _compute_bag_labels_logits_and_attn_maps(self, batch: Dict) -> Tuple[Tensor, Tensor, List]:
+        # The batch dict contains lists of tensors of different sizes, for all bags in the batch.
+        # This means we can't stack them along a new axis without padding to the same length.
+        # We could alternatively concatenate them, but this would require other changes (e.g. in
+        # the attention layers) to correctly split the tensors by bag/slide ID.
+        bag_labels_list = []
+        bag_logits_list = []
+        bag_attn_list = []
+        for bag_idx in range(len(batch[self.label_column])):
+            images = batch[TilesDataset.IMAGE_COLUMN][bag_idx]
+            labels = batch[self.label_column][bag_idx]
+            bag_labels_list.append(self.get_bag_label(labels))
+            logit, attn = self(images)
+            bag_logits_list.append(logit.view(-1))
+            bag_attn_list.append(attn)
+        bag_logits = torch.stack(bag_logits_list)
+        bag_labels = torch.stack(bag_labels_list).view(-1)
+        return bag_logits, bag_labels, bag_attn_list
+
+    def _update_results_with_data_specific_info(self, batch: dict, results: dict) -> None:
+        results.update({ResultsKey.SLIDE_ID: batch[TilesDataset.SLIDE_ID_COLUMN],
+                        ResultsKey.TILE_ID: batch[TilesDataset.TILE_ID_COLUMN],
+                        ResultsKey.IMAGE_PATH: batch[TilesDataset.PATH_COLUMN]})
+
+        if (TilesDataset.TILE_X_COLUMN in batch.keys()) and (TilesDataset.TILE_Y_COLUMN in batch.keys()):
+            results.update({ResultsKey.TILE_X: batch[TilesDataset.TILE_X_COLUMN],
+                           ResultsKey.TILE_Y: batch[TilesDataset.TILE_Y_COLUMN]}
+                           )
+        else:
+            if is_global_rank_zero():
```

Comment:
I think this could be done concisely with `from pytorch_lightning.utilities import rank_zero_warn`